server/sessionstore.go

```diff
@@ -21,6 +21,34 @@ import (
 	"github.com/tinode/chat/server/store/types"
 )
 
+type boundedWaitGroup struct {
+	wg  sync.WaitGroup
+	sem chan struct{}
+}
+
+func newBoundedWaitGroup(capacity int) *boundedWaitGroup {
+	return &boundedWaitGroup{sem: make(chan struct{}, capacity)}
+}
+
+func (w *boundedWaitGroup) Add(delta int) {
+	if delta <= 0 {
+		return
+	}
+	for i := 0; i < delta; i++ {
+		w.sem <- struct{}{}
+	}
+	w.wg.Add(delta)
+}
+
+func (w *boundedWaitGroup) Done() {
+	<-w.sem
```

Comment:
It looks like it would just block forever if the number of `Done()` exceeds the number of `Add()`. Maybe panic in such a case?